sstables/sstables.cc

```diff
@@ -2962,6 +2962,8 @@ sstable::unlink(storage::sync_dir sync) noexcept {
 
     co_await std::move(remove_fut);
     _stats.on_delete();
+    // Remove the sst from manager's reclaimed list to prevent any attempts to reload its components.
+    _manager._reclaimed.erase(*this);
 }
```

Comment:
if all unlink goes through manager, then we can avoid sstable not being aware of this implementation detail that belongs to manager (which lives in a layer above). see sstables_manager::delete_atomically(). or at least hide it like this: _manager.on_ulink(*_this). ideally the former, but I guess we can live with the latter.

note there can still be a race between reloader and unlink, if unlink happens after reloader picks sstable being unlinked, and before reloader starts parsing filter, but I guess that's benign as the sstable is not in the active sstable set anymore, and the failure to reload only means that sstable will live without a filter until its object is deleted, right?

Reply:
> if all unlink goes through manager, then we can avoid sstable not being aware of this implementation detail that belongs to manager (which lives in a layer above). see sstables_manager::delete_atomically(). or at least hide it like this: _manager.on_ulink(*_this). ideally the former, but I guess we can live with the latter.
> 

unlink had other callers not just the manager, so I have uses the second option (i.e.) `on_ulink`

> note there can still be a race between reloader and unlink, if unlink happens after reloader picks sstable being unlinked, and before reloader starts parsing filter, but I guess that's benign as the sstable is not in the active sstable set anymore, and the failure to reload only means that sstable will live without a filter until its object is deleted, right?

You are right, that will not be a problem.

Reply:
thanks